### PR TITLE
Use Handlebars Template for Benchmark CLI Output

### PR DESCRIPTION
**Pull request by shawntabrizi**
Supersedes https://github.com/paritytech/substrate/pull/7292
Thanks @zadmarbella for the initial start!

This PR changes how the final WeightInfo file is generated, basically completely re-writing the `writer`. Rather than printing custom lines into a file, we instead construct an object with the raw data, and allow a Handlebars template to parse that data to generate the final file.

This allows users to fully customize how they want to represent and organize their weight files. The objects generated are regular rust objects that can be easily extended if needed.

## Breaking Changes

* No more `--trait-def` flag to output a different "trait" file
* No more `--spaces` flag for using spaces
* No more `--struct` or `--trait` flag for renaming the struct/trait

## New Features

* New `--template` flag for specifying a handlebars template to use for the output generation. If not specified, the default template provided by Substrate will be used.
* Ability to specify just a folder or the full file name for benchmark output

The template provided with the frame-benchmarking-cli by default is good for creating weight files for downstream nodes, and I have provided in the `.maintain` folder a template good for building new pallets.

Within the handlebars template, you get access to all data compiled for you by the benchmarking cli:

```
// This is the final structure we will pass to the Handlebars template.
#[derive(Serialize, Default, Debug)]
struct TemplateData {
	args: Vec<String>,
	date: String,
	version: String,
	pallet: String,
	header: String,
	cmd: CmdData,
	benchmarks: Vec<BenchmarkData>,
}

// This was the final data we have about each benchmark.
#[derive(Serialize, Default, Debug, Clone)]
struct BenchmarkData {
	name: String,
	components: Vec<Component>,
	base_weight: u128,
	base_reads: u128,
	base_writes: u128,
	component_weight: Vec<ComponentSlope>,
	component_reads: Vec<ComponentSlope>,
	component_writes: Vec<ComponentSlope>,
}

// This forwards some specific metadata from the `BenchmarkCmd`
#[derive(Serialize, Default, Debug, Clone)]
struct CmdData {
	steps: Vec<u32>,
	repeat: u32,
	lowest_range_values: Vec<u32>,
	highest_range_values: Vec<u32>,
	execution: String,
	wasm_execution: String,
	chain: String,
	db_cache: u32,
}

// This encodes the component name and whether that component is used.
#[derive(Serialize, Debug, Clone, Eq, PartialEq)]
struct Component {
	name: String,
	is_used: bool,
}

// This encodes the slope of some benchmark related to a component.
#[derive(Serialize, Debug, Clone, Eq, PartialEq)]
struct ComponentSlope {
	name: String,
	slope: u128,
}
```

You can see the included template in this PR to see how to access that data.

### CODE REVIEW

**Review by thiolliere — APPROVED** [review]
looks good to me,
do you have an example of output with this template ?

**Comment by shawntabrizi** [comment]
@thiolliere example output can be found here: https://github.com/paritytech/substrate/pull/7391

File: .maintain/frame-weight-template.hbs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+//! DATE: {{date}}, STEPS: {{cmd.steps}}, REPEAT: {{cmd.repeat}}, LOW RANGE: {{cmd.lowest_range_values}}, HIGH RANGE: {{cmd.highest_range_values}}
+//! EXECUTION: {{cmd.execution}}, WASM-EXECUTION: {{cmd.wasm_execution}}, CHAIN: {{cmd.chain}}, DB CACHE: {{cmd.db_cache}}
+
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+
+use frame_support::{traits::Get, weights::Weight};
+use sp_std::marker::PhantomData;
+
+pub trait WeightInfo {
```
Comment (thiolliere) [inline .maintain/frame-weight-template.hbs:30]: Should have docs

File: .maintain/frame-weight-template.hbs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	{{#each benchmarks as |benchmark| ~}}
+	fn {{benchmark.name~}}
+	(
+		{{~#each benchmark.components as |c| ~}}
+		_{{c.name}}: u32, {{/each~}}
+	) -> Weight;
+	{{/each}}
+}
+
+pub struct SubstrateWeight<T>(PhantomData<T>);
```
Comment (thiolliere) [inline .maintain/frame-weight-template.hbs:40]: should also have docs

File: utils/frame/benchmarking-cli/src/template.hbs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+//! DATE: {{date}}, STEPS: {{cmd.steps}}, REPEAT: {{cmd.repeat}}, LOW RANGE: {{cmd.lowest_range_values}}, HIGH RANGE: {{cmd.highest_range_values}}
+//! EXECUTION: {{cmd.execution}}, WASM-EXECUTION: {{cmd.wasm_execution}}, CHAIN: {{cmd.chain}}, DB CACHE: {{cmd.db_cache}}
+
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+
+use frame_support::{traits::Get, weights::Weight};
+use sp_std::marker::PhantomData;
+
+pub struct WeightInfo<T>(PhantomData<T>);
```
Comment (thiolliere) [inline utils/frame/benchmarking-cli/src/template.hbs:14]: also better to have doc here.

**Review by thiolliere — COMMENTED** [review]
(some additional doc needed, but still code is good)

File: utils/frame/benchmarking-cli/src/writer.rs [hunk 4]
```diff
@@ -17,314 +17,401 @@
...
 ) -> Result<(), std::io::Error> {
+	// Use custom template if provided.
+	let template: String = match &cmd.template {
+		Some(template_file) => {
+			fs::read_to_string(template_file)?
+		},
+		None => {
+			TEMPLATE.to_string()
+		},
+	};
```
Comment (athei) [inline utils/frame/benchmarking-cli/src/writer.rs:206]: I think it better to use the default value feature of structopt than this custom logic. By doing that the default template is automatically discoverable on the CLI.

File: utils/frame/benchmarking-cli/src/lib.rs [hunk 5]
```diff
@@ -68,9 +68,9 @@ pub struct BenchmarkCmd {
 	#[structopt(long)]
 	pub header: Option<std::path::PathBuf>,
 
-	/// Output the trait definition to a Rust file.
+	/// Path to Handlebars template file used for outputting benchmark results. (Optional)
 	#[structopt(long)]
-	pub trait_def: bool,
+	pub template: Option<std::path::PathBuf>,
```
Comment (athei) [inline utils/frame/benchmarking-cli/src/lib.rs:73]: Shouldn't we the default value feature here to make the default template discoverable on the CLI?

**Review by athei — COMMENTED** [review]
Why do we have two different templates? Why can't the FRAME template be the default?

**Comment by shawntabrizi** [comment]
> Why do we have two different templates? Why can't the FRAME template be the default?

The output when you are building a weights.rs file for a custom pallet versus generating all the weights for your runtime are different.

For example, in my proposal, the weights for a new pallet will contain the definition for the WeightInfo trait.

However, the default template should not necessarily be for new pallet developers, but simply the template needed to get you Substrate chain running.

**Comment by athei** [comment]
Does that mean that the `WeightInfo` trait does no longer live inside the pallet itself for substrate pallets?

**Comment by shawntabrizi** [comment]
@athei yes, that is the plan. look here https://github.com/paritytech/substrate/pull/7391 and give feedback

**Review by athei — APPROVED** [review]

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**